README.md
# MMM-HK-Transport
<B>Station monitor</B> for the <B>Hong Kong Transport</B>.<P>

This module is an extension of the amazing [MagicMirror<sup>2</sup>](https://github.com/MichMich/MagicMirror) project by [MichMich](https://github.com/MichMich/) which has inspired me to share my coding skills with others as well. Check it out, you know you want to :). <P>

It's always nice to see mirrors using my work, so feel free to send me some screenshots of your implementations.<P>

Lastly, why not join in on our discussions at the official [MagicMirror<sup>2</sup> Forum](http://forum.magicmirror.builders/)?

## Screenshots

`en`             |  `zh-tw` or `zh-hk`
:-------------------------:|:-------------------------:
![English version](screenshots/screenshot_en.png)  |  ![Chinese version](screenshots/screenshot_zh.png) |

## Current version

v1.3.0

## Languages
As of version 1.3.0, MMM-HK-Transport features language support for `Chinese (zh-hk and zh-tw)` and `English (en)` mirrors.

## Prerequisite
A working installation of [MagicMirror<sup>2</sup>](https://github.com/MichMich/MagicMirror)
 
## Dependencies
  * npm
  * [got](https://github.com/sindresorhus/got)

## Installation
Go to MagicMirror folder and execute the following command:
```
cd modules
git clone https://github.com/winstonma/MMM-HK-Transport.git
cd MMM-HK-Transport
npm install
```

## Module behavior
Please note that this module auto-creates a module header which displays the name of the chosen Hong Kong local transport stop. It is therefore recommended not to add a 'header' entry to your config.js for this module.<P>
This module automatically disappears from your mirror as soon as a station has stopped offering connections at night. It reappears as soon as your chosen station is scheduled to be served again.<P>
This module has been programmed to allow for multiple instances. Simply add more MMM-HK-Transport config entries to your config.js file to display multiple stations and configure them according to your needs.

## Configuration
Sample minimum configuration entry for your `~/MagicMirror/config/config.js`:

    ...
    
    {
        module: 'MMM-HK-Transport',
        position: 'top_left',
        config: {
            stops: [
                {
                    stopID: 'HKStop_KowloonCentralPostOffice_N_1'
                }
            ]
        }
    } 				// If this isn't your last module, add a comma after the bracket
    
    ...

Sample configuration entry for your `~/MagicMirror/config/config.js` with optional parameters:

    ...
    
    {
        module: 'MMM-HK-Transport',
        position: 'top_left',
        config: {
            stopID: 'HKStop_KowloonCentralPostOffice_N_1',
            showLabelRow: true, // Show or hide column headers
            reloadInterval: 60000 	// How often should the information be updated? (In milliseconds)
        }
    } 						// If this isn't your last module, add a comma after the bracket
    
    ...

## Figuring out the correct stopID
1. Open your web browser and navigate to the [CityMapper Hong Kong Web Page](https://citymapper.com/hong-kong).
2. Go to `LINES` search box enter the bus line number.
3. Once you can see the stop in your browser, click the stop.
4. When a new page is being displayed, check the link (e.g. https://citymapper.com/hong-kong/bus/stops/HKStop_KowloonCentralPostOffice_N_1). Note the last portion of the link (e.g. `HKStop_KowloonCentralPostOffice_N_1`)is the `StopID` you are looking for.

## Config Options
| **Option** | **Default** | **Description** |
| :---: | :---: | --- |
| stopID | HKStop_KowloonCentralPostOffice_N_1 | <BR>Which stop would you like to have displayed? <BR><EM> Default: HKStop_KowloonCentralPostOffice_N_1</EM><P> |
| showLabelRow<BR>`optional` | true | <BR> Show or hide column headers<BR> <EM>Possible values: true, false</EM><P> |
| reloadInterval<BR>`optional`  | 60000 | <BR> How often should the information be updated? (In milliseconds) <BR><EM> Default: Every minute </EM><P> |


changelog.md
# MMM-HK-Transport changelog
This module adheres to Winston

## [1.3.0] - 2020-10-02
* Update CityMapper API
* Support `zh-hk`
* Support Red/Green Van, Tram stops
* Time display follows the `timeFormat` parameter in `config.js`
* Stop name will be collected from the API
* Replace the API fetching module
* Fix route with multiple services not being displayed

## [1.2.0] - 2020-05-08
* Some route is not running during display, so it would be disabled
* Fix some stop name display bug

## [1.1.0] - 2017-04-07
* Add support for multiple stops

## [1.0.0] - 2017-03-24 
* Initial release


fetcher.js
/* Magic Mirror
 * Module: MMM-HK-Transport
 *
 * By Winston / https://github.com/winstonma
 * MIT Licensed.
 */

const got = require('got');

/* Fetcher
 * Responsible for requesting an update on the set interval and broadcasting the data.
 *
 * attribute url string - URL of the news feed.
 * attribute reloadInterval number - Reload interval in milliseconds.
 */

var Fetcher = function (url, reloadInterval) {
    var self = this;
    if (reloadInterval < 1000) {
        reloadInterval = 1000;
    }

    var reloadTimer = null;
    var items = null;

    var fetchFailedCallback = function () { };
    var itemsReceivedCallback = function () { };

    /* private methods */

    /* fetchETAs()
     * Request the new items.
     */

    var fetchETAs = function () {
        clearTimeout(reloadTimer);
        reloadTimer = null;
        items = null;

        (async () => {
            try {
                const response = await got(url, {
                    responseType: 'json'
                });
                items = response.body;
                self.broadcastItems();
                scheduleTimer();
            } catch (error) {
                console.log(error.response.body);
                fetchFailedCallback(self, error);
                scheduleTimer();
            }
        })();
    };

    /* scheduleTimer()
     * Schedule the timer for the next update.
     */

    var scheduleTimer = function () {
        //console.log('Schedule update timer.');
        clearTimeout(reloadTimer);
        reloadTimer = setTimeout(function () {
            fetchETAs();
        }, reloadInterval);
    };

    /* public methods */

    /* setReloadInterval()
     * Update the reload interval, but only if we need to increase the speed.
     *
     * attribute interval number - Interval for the update in milliseconds.
     */
    this.setReloadInterval = function (interval) {
        if (interval > 1000 && interval < reloadInterval) {
            reloadInterval = interval;
        }
    };

    /* startFetch()
     * Initiate fetchETAs();
     */
    this.startFetch = function () {
        fetchETAs();
    };

    /* broadcastItems()
     * Broadcast the existing items.
     */
    this.broadcastItems = function () {
        if (typeof items === "undefined") {
            //console.log('No items to broadcast yet.');
            return;
        }
        //console.log('Broadcasting ' + items.length + ' items.');
        itemsReceivedCallback(self);
    };

    this.onReceive = function (callback) {
        itemsReceivedCallback = callback;
    };

    this.onError = function (callback) {
        fetchFailedCallback = callback;
    };

    this.url = function () {
        return url;
    };

    this.items = function () {
        return items;
    };
};

module.exports = Fetcher;

etafetcher.js
/* Magic Mirror
 * Node Helper: MMM-HK-Transport - ETAFetcher
 *
 * By Winston / https://github.com/winstonma
 * AGPL-3.0 Licensed.
 */
const Log = require("../../js/logger.js");
const got = require('got');

/**
 * Responsible for requesting an update on the set interval and broadcasting the data.
 *
 * @param {string} url URL of the ETA feed.
 * @param {string} stopID stop ID of the stop
 * @param {number} reloadInterval Reload interval in milliseconds.
 * @class
 */
const ETAFetcher = function (url, stopID, reloadInterval) {
    const self = this;

    let reloadTimer = null;
    let item = null;

    let fetchFailedCallback = function () { };
    let itemsReceivedCallback = function () { };

    if (reloadInterval < 1000) {
        reloadInterval = 1000;
    }

    /* private methods */

    /**
     * Request the ETA
     */
    const fetchETAs = function () {
        clearTimeout(reloadTimer);
        reloadTimer = null;
        item = [];

        (async () => {
            try {
                const {body} = await got(url, {
                    responseType: 'json'
                });
                item = body;
                self.broadcastItems();
                scheduleTimer();
            } catch (error) {
                console.log(error.response.body);
                fetchFailedCallback(self, error);
                scheduleTimer();
            }
        })();
    };

	/**
	 * Schedule the timer for the next update.
	 */
	const scheduleTimer = function () {
		clearTimeout(reloadTimer);
		reloadTimer = setTimeout(function () {
			fetchETAs();
		}, reloadInterval);
	};

    /* public methods */

	/**
	 * Update the reload interval, but only if we need to increase the speed.
	 *
	 * @param {number} interval Interval for the update in milliseconds.
	 */
	this.setReloadInterval = function (interval) {
		if (interval > 1000 && interval < reloadInterval) {
			reloadInterval = interval;
		}
	};

	/**
	 * Initiate fetchETAs();
	 */
    this.startFetch = function () {
        fetchETAs();
    };

	/**
	 * Broadcast the existing item.
	 */
    this.broadcastItems = function () {
		if (item.length <= 0) {
			Log.info("ETA-Fetcher: No item to broadcast yet.");
			return;
        }
        Log.info(`ETA-Fetcher: Broadcasting item for stop ID ${stopID}`);
		itemsReceivedCallback(self);
    };

	this.onReceive = function (callback) {
		itemsReceivedCallback = callback;
	};

	this.onError = function (callback) {
		fetchFailedCallback = callback;
	};

    this.url = function () {
        return url;
    };

    this.items = function () {
        return item;
    };
};

module.exports = ETAFetcher;

node_helper.js
/* Magic Mirror
 * Node Helper: MMM-HK-Transport
 *
 * By Winston / https://github.com/winstonma
 * AGPL-3.0 Licensed.
 */

const validUrl = require("valid-url");
const ETAFetcher = require("./etafetcher.js");
const NodeHelper = require("node_helper");
const got = require('got');
const Log = require("../../js/logger");

module.exports = NodeHelper.create({
    // Override start method.
    start: function () {
        Log.log("Starting node helper for: " + this.name);
        this.fetchers = [];
    },

    // Override socketNotificationReceived received.
    socketNotificationReceived: function (notification, payload) {
        if (notification === 'ADD_STOP') {
            this.getStopInfo(payload.stop, payload.config);
            return;
        }
    },

    /**
     * Obtain the stop id from the URL
     *
     * @param {object} stopInfo The stopInfo.
     * @param {object} config The configuration object.
     */
    getStopInfo: function (stopInfo, config) {
        const baseURL = "https://citymapper.com/api/3/stopinfo?region_id=hk-hongkong&ids=";

        const url = baseURL + stopInfo.stopID;
        (async () => {
            try {
                const {body} = await got(url, {
                    responseType: 'json'
                });
                this.createFetcher(body.stops[0].id, stopInfo, config);
            } catch (error) {
                this.sendSocketNotification("FETCH_ERROR", {
                    stopID: stopInfo.stopID,
                    error: error
                });
            }
        })();
    },

    /**
     * Creates a fetcher for a ETA feed if it doesn't exist yet.
     * Otherwise it reuses the existing one.
     *
     * @param {object} stopID The stopID.
     * @param {object} stopInfo The stopInfo.
     * @param {object} config The configuration object.
     */
    createFetcher: function (stopID, stopInfo, config) {
        const url = config.cityMapperURL + stopID || "";
        const reloadInterval = stopInfo.reloadInterval || config.reloadInterval || 5 * 60 * 1000;

        if (!validUrl.isUri(url)) {
            this.sendSocketNotification("INCORRECT_URL", url);
            return;
        }

        let fetcher;
        if (typeof this.fetchers[stopInfo.stopID] === "undefined") {
            Log.log("Create new CityMapper fetcher for url: " + url + " - Interval: " + reloadInterval);
            fetcher = new ETAFetcher(url, stopInfo.stopID, reloadInterval);

            fetcher.onReceive(() => {
				this.broadcastFeeds();
			});

            fetcher.onError((fetcher, error) => {
                this.sendSocketNotification("FETCH_ERROR", {
                    url: fetcher.url(),
                    error: error
                });
            });

            this.fetchers[stopInfo.stopID] = fetcher;
        } else {
            Log.log("Use existing CityMapper fetcher for url: " + url);
            fetcher = this.fetchers[stopInfo.stopID];
            fetcher.setReloadInterval(reloadInterval);
            fetcher.broadcastItems();
        }

        fetcher.startFetch();
    },

	/**
	 * Creates an object with all ETA items of the different registered stopID,
	 * and broadcasts these using sendSocketNotification.
	 */
    broadcastFeeds: function () {
        const feeds = Object.entries(this.fetchers)
            .map(([key, fetcher]) => [key, fetcher.items()])
            .reduce((r, [k, v]) => ({ ...r, [k]: v }), {});

        this.sendSocketNotification("STOP_ITEMS", feeds);
    },
});

MMM-HK-Transport.js
/* Magic Mirror
 * Module: MMM-HK-Transport
 *
 * By Winston / https://github.com/winstonma
 * AGPL-3.0 Licensed.
 */

Module.register("MMM-HK-Transport", {

    defaults: {
        stops: [
            {
                stopID: 'HKStop_KowloonCentralPostOffice_N_1'
            }
        ],
        timeFormat: (config.timeFormat !== 24) ? "h:mm" : "HH:mm",
        showLabelRow: true,
        cityMapperURL: 'https://citymapper.com/api/1/departures?headways=1&region_id=hk-hongkong&ids=',
        reloadInterval: 1 * 60 * 1000       // every minute
    },

    // Define required scripts.
    getScripts: function () {
        return ["moment.js"];
    },

    getTranslations: function () {
        return {
            "en": "translations/en.json",
            "zh-hk": "translations/zh.json",
            "zh-tw": "translations/zh.json"
        };
    },

    getStyles: function () {
        return ["MMM-HK-Transport.css", "font-awesome.css"];
    },

    start: function () {
        Log.info("Starting module: " + this.name);

        this.cityMapperData = {};

        this.registerStops();
    },

    /**
     * Registers the stops to be used by the backend.
     */
    registerStops: function () {
        this.config.stops.forEach(stop => {
            this.sendSocketNotification("ADD_STOP", {
                stop: stop,
                config: this.config
            });
        });
    },

    /**
    * Generate an ordered list of items for this configured module.
    *
    * @param {object} etas An object with ETAs returned by the node helper.
    */
    generateETA: function (etas) {
        this.cityMapperData = Object.entries(etas)
            .filter(([stopID,]) => this.subscribedToETA(stopID))
            .map(([k, v]) => {
                const stop = v.stops[0];

                // Merge sevices and routes into one
                const stopInfo = stop.services.map(service => {
                    return {
                        route: stop.routes.find(element => element.id == service.route_id),
                        service: service
                    }
                }).sort((a, b) => (a.route.id > b.route.id) ? 1 : -1);

                stop.stopInfo = stopInfo;
                delete stop.services;
                delete stop.routes;

                return [k, v];
            })
            .reduce((r, [k, v]) => ({ ...r, [k]: v }), {});
    },

    /**
     * Check if this module is configured to show this ETA.
     *
     * @param {string} stopID stopID to check.
     * @returns {boolean} True if it is subscribed, false otherwise
     */
    subscribedToETA: function (stopID) {
        return this.config.stops.some(stop => stop.stopID === stopID);
    },

    // Override socket notification handler.
    socketNotificationReceived: function (notification, payload) {
        if (notification === "STOP_ITEMS") {
            this.generateETA(payload);
            this.updateDom();
        }
    },

    // Override dom generator.
    getDom: function () {
        const wrapper = document.createElement("div");

        if (Object.keys(this.cityMapperData).length === 0) {
            wrapper.appendChild(this.createStopHeader(null));
            let text = document.createElement("div");
            text.innerHTML = this.translate("LOADING");
            text.className = "small dimmed";
            wrapper.appendChild(text);
            return wrapper;
        }

        Object.entries(this.cityMapperData).forEach(([stopID, stop]) => {
            wrapper.appendChild(this.createStopHeader(stop.stops[0]));
            const stopConfig = this.config.stops.find(stop => stop.stopID == stopID);
            wrapper.appendChild(this.createStops(stopConfig, stop.stops[0]));
        });

        return wrapper;
    },

    getDisplayString: function (input) {
        const langTable = {
            'zh-tw': 'zh',
            'zh-hk': 'zh',
            'zh-cn': 'zh'
        }

        const REGEX_CHINESE = /[\u4e00-\u9fff]|[\u3400-\u4dbf]|[\u{20000}-\u{2a6df}]|[\u{2a700}-\u{2b73f}]|[\u{2b740}-\u{2b81f}]|[\u{2b820}-\u{2ceaf}]|[\uf900-\ufaff]|[\u3300-\u33ff]|[\ufe30-\ufe4f]|[\uf900-\ufaff]|[\u{2f800}-\u{2fa1f}]/u;
        const splitStr = input.split(' ');
        const chiWords = splitStr.filter((string) => REGEX_CHINESE.test(string));
        const engWords = splitStr.filter((string) => !REGEX_CHINESE.test(string)).join(' ');

        if (langTable[config.language] && chiWords.length)
            return chiWords;
        return engWords;
    },

    createStopHeader: function (stop) {
        // Auto-create MagicMirror header
        let header = document.createElement("header");
        header.innerHTML = (stop == null) ? this.name : this.getDisplayString(stop.name);
        return header;
    },

    createStops: function (stopConfig, stop) {
        // Start creating connections table
        let table = document.createElement("table");
        table.classList.add("small", "table");
        table.border = '0';

        const showLabelRow = (typeof stopConfig.showLabelRow !== 'undefined') ? stopConfig.showLabelRow : this.config.showLabelRow;

        if (showLabelRow) {
            table.appendChild(this.createLabelRow());
        }

        table.appendChild(this.createSpacerRow());

        // This loop create the table that display the content
        stop.stopInfo.forEach(element => {
            const rowContent = this.createDataRow(element);
            if (rowContent)
                table.appendChild(rowContent);
        });

        return table;

    },

    createLabelRow: function () {
        let labelRow = document.createElement("tr");

        let lineLabel = document.createElement("th");
        lineLabel.className = "line";
        lineLabel.innerHTML = this.translate("LINE");
        labelRow.appendChild(lineLabel);

        let destinationLabel = document.createElement("th");
        destinationLabel.className = "destination";
        destinationLabel.innerHTML = this.translate("DESTINATION");
        labelRow.appendChild(destinationLabel);

        let departureLabel = document.createElement("th");
        departureLabel.className = "departure";
        departureLabel.innerHTML = this.translate("DEPARTURE");
        labelRow.appendChild(departureLabel);

        return labelRow;
    },

    createSpacerRow: function () {
        let spacerRow = document.createElement("tr");

        let spacerHeader = document.createElement("th");
        spacerHeader.className = "spacerRow";
        spacerHeader.setAttribute("colSpan", "3");
        spacerHeader.innerHTML = "";
        spacerRow.appendChild(spacerHeader);

        return spacerRow;
    },

    createNoTramRow: function () {
        let noTramRow = document.createElement("tr");

        let noTramHeader = document.createElement("th");
        noTramHeader.className = "noTramRow";
        noTramHeader.setAttribute("colSpan", "3");
        noTramHeader.innerHTML = this.translate("NO-TRAMS");
        noTramRow.appendChild(noTramHeader);

        return noTramRow;
    },

    createDataRow: function (routeObj) {
        let etaArray;

        if (routeObj.service.next_departures) {
            const hoursFromNow = moment.duration(moment(routeObj.service.next_departures[0]).diff(moment()))
                .asHours();
            if (hoursFromNow < 1) {
                etaArray = routeObj.service.next_departures.map(etaStr => moment(etaStr).format(this.config.timeFormat));
            }
        } else if (routeObj.service.headway_seconds_range) {
            const [rangeBottom, rangeTop] = routeObj.service.headway_seconds_range.map(seconds => Math.floor(seconds / 60));
            const midStr = (rangeBottom == rangeTop) ? rangeBottom : `${rangeBottom}—${rangeTop}`;
            etaArray = this.translate("EVERY") + midStr + this.translate("MINUTES");
        } else if (routeObj.service.live_departures_seconds) {
            etaArray = routeObj.service.live_departures_seconds.map(seconds => moment().seconds(seconds).format(this.config.timeFormat));
        }

        if (!etaArray)
            return null;

        let row = document.createElement("tr");

        let line = document.createElement("td");
        line.className = "line";
        line.innerHTML = routeObj.route.name;
        if (routeObj.route.brand === "GMBBus")
            line.innerHTML += '<sup><i class="fas fa-shuttle-van"></i></sup>';
        row.appendChild(line);

        let destination = document.createElement("td");
        destination.className = "destination";
        destination.innerHTML = this.getDisplayString(routeObj.service.headsign);
        row.appendChild(destination);

        let departure = document.createElement("td");
        departure.className = "departure";
        departure.innerHTML = etaArray.toString();
        row.appendChild(departure);

        return row;
    }

});